documentation/specs/src/base-ledger/replay-protection.md

```diff
@@ -0,0 +1,100 @@
+# Replay Protection
+
+Namada supports a replay protection mechanism to prevent the execution of already processed transactions.
+
+## Tendermint
+
+[Tendermint]((https://docs.tendermint.com/v0.33/app-dev/app-development.html#replay-protection)) provides a first layer of protection against replay attacks in its mempool. The mechanism is based on a cache of previously seen transactions. This check though, like all the checks performed in `CheckTx`, is weak, since a malicious validator could always propose a block containing invalid transactions. There's therefore the need for a more robust replay protection mechanism implemented directly in the application.
+
+## WrapperTx
+
+`WrapperTx` is the only type of transaction currently accepted by the ledger. It must be protected from replay attacks because, if it wasn't, a malicious user could replay the transaction as is. Even if the inner transaction implemented replay protection or, for any reason, wasn't accepted, the signer of the wrapper would still pay for gas and fees, effectively suffering economic damage.
+
+To protect this transaction we can implement an in-protocol mechanism that requires us to expand the current definition of the struct to include a transaction counter:
+
+```rust
+pub struct WrapperTx {
+    /// The fee to be payed for including the tx
+    pub fee: Fee,
+    /// Used to determine an implicit account of the fee payer
+    pub pk: common::PublicKey,
+    /// The epoch in which the tx is to be submitted. This determines
+    /// which decryption key will be used
+    pub epoch: Epoch,
+    /// Max amount of gas that can be used when executing the inner tx
+    pub gas_limit: GasLimit,
+    /// Transaction counter for replay-protection
+    pub tx_counter: u32,
+    /// the encrypted payload
+    pub inner_tx: EncryptedTx,
+    /// sha-2 hash of the inner transaction acting as a commitment
+    /// the contents of the encrypted payload
+    pub tx_hash: Hash,
+}
+``` 
+
+In addition, we need a counter in the storage subspace of every implict address:
+
+```
+/$Address/tx_counter: u32
+```
+
+In `process_proposal` we check that `tx_counter` field in `WrapperTx` is greater or equal to the value currently in storage; if this doesn't hold, the transaction is considered invalid.
```

Comment:
I think the check here has to be for strict equality, because we only increment it by one - if we allow greater than the current value, that could still be replayed, because even after its updated, it would still be at least equal to it. We'd need to increment it to the highest counter value from all txs in a block, but that's still affected by the other issues you describe in batching and it's tricky with the wrapper/inner tx counter interplay, so I think there will be better options for tx batching.